docs/undo-and-time-travel.md

```diff
@@ -0,0 +1,40 @@
+# Undo and Time Travel for Tile Developers
+
+When a tile is used in a CLUE document, the history of changes to that tile's content model are recorded by the history framework. The details of the implementation of the framework are described in [history-framework.md](./history-framework.md). These history entries are used to support undo and time traveling.
+
+The history entries are based on the actions called on the tile's content model. If one action (the parent) calls another action (a child), a single entry is recorded with a name of the parent action.  This single entry has the changes from the parent and child actions.
+
+## withoutUndo
+
+If you want an action to be part of the time travel history, but not part of the undo stack you can use `withoutUndo()`. 
+
+The convention is to call this at the top of an action like this:
+
+```
+actions(self => {
+  doSomething() {
+    withoutUndo();
+    // make some changes to self or its children
+  }
+})
+```
+
+Currently if you add `withoutUndo` to an action and then you call this action from another action, the `withoutUndo` will be ignored. A warning message will be printed in this case describing how we can change the code to handle this. You can see that in [tree-monitor.ts](../src/models/history/tree-monitor.ts#L306)
+
+## Grouping Changes
+
+You will want to group changes into a single entry in some cases. Otherwise a user will have to undo multiple times when they'd expect to only undo once.
+
+There are currently 2 ways to group changes:
+- use a parent action
+- use a MST `flow`
```

Comment:
Would it be possible to support a third mechanism in which changes can be grouped after the fact? I ask because one way I've seen keystrokes handled for undo purposes in text editors is to have the first keystroke add an undoable action to the history and then subsequent keystrokes are grouped/merged with the first keystroke. Thus, the first non-keystroke action is what ends the undoable editing session. This allows model changes to be propagated to collaborators immediately while still allowing grouping of keystrokes into larger chunks for undo. From an implementation standpoint, it looks something like this (pseudocode):
```javascript
addHistoryEntry(action) {
  if (canMerge(action, lastAction()) {
    merge(action, lastAction())
  }
  else {
    appendAction(action)
  }
}
```

Reply:
It is possible, but I'm not sure the right mechanism to do it yet. I'm also not sure about the implications.

Currently history entries are marked as complete when the parent action and any residual cascading shared model updates are finished. In the model where the history entries are how documents on other machines stay synchronized, these complete entries would be broadcast when they are complete. That approach would be broken if we start modifying them after they are sent. But I remember this was a feature of that collaborative state framework we were looking at before, so we can probably come up with an approach where each individual machine can do this merging as the second mergable event comes in. 

Or alternatively we can record these as individual events and it is just in the undo stack where they get grouped.

Regarding the mechanism, we'd need a way for code within an action to decide if it can be merged with the last action (or possibly one before the last action action). So that could look like this:
```
.actions(self => ({
  doSomething() {
    if(getHistory().last.name === 'doSomething') {
      mergeCurrentActionWith(getHistory().last);
    }
  }
}))
```
The condition will have to be more complex than this because we'd need to check if the last action is from the same model.

On the other side, the flow construct is kind of similar. But probably it won't work well enough. I haven't tested it, but I believe any of the changes made during a flow are instantly available in the document tree. I think they result in `onPatch` being called immediately. The thing that is delayed is the "transaction". So anything observing the changed properties will not be updated until the transaction/flow is finished. This way the tree is supposed to be in a consistent state before these observers are updated.  So in theory we could transmit the document using onPatch, but we might send inconsistent documents.